pkg/pillar/cmd/zedrouter/flowstats.go

```diff
@@ -0,0 +1,706 @@
+				}
+			}
+			// flow record done for the bridge/app
+			//flowdata.CastFlowStatus
+			flowKey := scope.UUID.String() + scope.NetUUID.String()
+			pub.Publish(flowKey, &flowdata)
+		}
+	}
+
+	// remove the dns data already uploaded
+	if dnsPacked {
+		for bnx := range instData.bnNet {
+			bnNum, err := bridgeStrToNum(bnx)
+			if err != nil {
+				continue
+			}
+			dnssys[bnNum].Snoop = nil
+			log.Infof("!!FlowStats: clear dns record for bn%d", bnNum)
+		}
+	}
+}
+
+// conntrack flow of two uni-directional stats into one
+// bireditional flow stats
+func flowMergeProcess(entry *netlink.ConntrackFlow, instData networkAttrs) flowStats {
+	var ipFlow flowStats
+	var forwSrcApp, forwDstApp, backSrcApp, backDstApp bool
+	var AppNum int
+
+	// for the current phase of implementation, we ignore the flowstats which has not timed out yet
+	ipflowTimeOut := int32(entry.TimeOut)
+	if ipflowTimeOut > timeoutSec {
+		return ipFlow
+	}
+
+	ipFlow.aclNum = entry.Mark & markMask
+	ipFlow.appNum = uint8(entry.Mark >> appShiftBits)
+	AppNum = int(ipFlow.appNum)
+
+	//AppNum := 1 // XXX hack, = ipFlow.appNum later
+	//ipFlow.appNum = uint8(AppNum)
+	ipFlow.TimeStart = entry.TimeStart
+	ipFlow.TimeStop = entry.TimeStop
+	ipFlow.TimeOut = entry.TimeOut
+	ipFlow.Proto = entry.Forward.Protocol
+	ipFlow.IsTimeOut = true
+
+	// Assume the App has an assigned IP address(es) first
+	if len(instData.appIPAddrs[AppNum]) > 0 {
+		ipFlow.dbg1 = 1
+		forwSrcApp = checkAppIPAddr(instData.appIPAddrs[AppNum], entry.Forward.SrcIP)
+		if forwSrcApp == false {
+			ipFlow.dbg1 = 2
+			forwDstApp = checkAppIPAddr(instData.appIPAddrs[AppNum], entry.Forward.DstIP)
+			if forwDstApp == false {
+				ipFlow.dbg1 = 3
+				backSrcApp = checkAppIPAddr(instData.appIPAddrs[AppNum], entry.Reverse.SrcIP)
+				if backSrcApp == false {
+					ipFlow.dbg1 = 4
+					backDstApp = checkAppIPAddr(instData.appIPAddrs[AppNum], entry.Reverse.DstIP)
+				} else {
+					ipFlow.dbg1 = 5 // XXX
+				}
+			}
+		}
+	}
+
+	if forwSrcApp {
+		ipFlow.dbg2 = 1
+		ipFlow.SrcIP = entry.Forward.SrcIP
+		ipFlow.DstIP = entry.Reverse.SrcIP
+		ipFlow.SrcPort = entry.Forward.SrcPort
+		ipFlow.DstPort = entry.Reverse.DstPort
+		ipFlow.AppInitiate = true
+	} else if forwDstApp {
+		ipFlow.dbg2 = 2
+		ipFlow.DstIP = entry.Forward.DstIP
+		ipFlow.SrcIP = entry.Reverse.DstIP
+		ipFlow.DstPort = entry.Forward.DstPort
+		ipFlow.SrcPort = entry.Reverse.DstPort
+	} else if backSrcApp {
+		ipFlow.dbg2 = 3
+		ipFlow.SrcIP = entry.Forward.SrcIP
+		ipFlow.DstIP = entry.Reverse.SrcIP
+		ipFlow.SrcPort = entry.Forward.SrcPort
+		ipFlow.DstPort = entry.Reverse.SrcPort
+	} else if backDstApp {
+		ipFlow.dbg2 = 4
+		ipFlow.SrcIP = entry.Reverse.DstIP
+		ipFlow.DstIP = entry.Forward.DstIP
+		ipFlow.SrcPort = entry.Reverse.DstPort
+		ipFlow.DstPort = entry.Forward.DstPort
+		ipFlow.AppInitiate = true
+	} else { // if we can not find our App endpoint is part of the flow, something is wrong
```

Comment:
This is a valid case for switch network instance. We do not know the IP addresses of apps that are attached to switch network instances. Should think about handling this case.

Reply:
right. for now, i'll move this part a little bit, so, if we don't find the app, then do this search through finding external networks, and reverse finding the app ip address